modules/packages/Futures.chpl

```diff
@@ -0,0 +1,240 @@
+
+   writeln(F.get() == (3 * ((2 * X) + 7))); // prints "true"
+
+.. code-block:: chapel
+
+   use Futures;
+
+   config const X = 1234;
+
+   var F = async(lambda(x:int) { return x:string; }, X)
+     .andThen(lambda(x:string) { return(x.length); });
+
+   writeln(F.get()); // prints "4"
+
+ */
+
+module Futures {
+
+  use Reflection;
+  use RefCount;
+
+  pragma "no doc"
+  class FutureClass: RefCountBase {
+
+    type retType;
+
+    var value: single retType;
+
+    proc FutureClass(type retType) {
+    }
+
+  } // class FutureClass
+
+  /*
+    A container that can store the result of an asynchronous operation,
+    which can be retrieved when the result is ready.
+
+    A future is not created directly. Instead, one is created by calling the
+    :proc:`async()` function or the :proc:`Future.andThen()` method on
+    an already-existing future.
+   */
+  record Future {
+
+    /*
+      The return type of the future.
+     */
+    type retType;
+
+    pragma "no doc"
+    var classRef: FutureClass(retType);
+
+    pragma "no doc"
+    proc Future(type retType) {
+      acquire(new FutureClass(retType));
+    }
+
+    pragma "no doc"
+    proc ~Future() {
+      release();
+    }
+
+    /*
+      Get the result of a future, blocking until it is available.
+     */
+    proc get(): retType {
+      return classRef.value.readFF();
+    }
+
+    /*
+      Test whether the result of the future is available.
+     */
+    proc isReady(): bool {
+      return classRef.value.isFull;
+    }
+
+    /*
+      Asynchronously execute a function as a continuation of the future.
+
+      The function argument `taskFn` must take a single argument of type
+      `retType` (i.e., the return type of the parent future) and will be
+      executed when the parent future's value is available.
+
+      :arg taskFn: The function to invoke as a continuation.
+      :returns: A future of the return type of `taskFn`
+     */
+    proc andThen(taskFn) {
+      /*
+      if !canResolveMethod(taskFn, "this", retType) then
+        compilerError("Mismatch in async task function and arguments");
+      */
+      var f: Future(taskFn.retType);
+      begin f.classRef.value.writeEF(taskFn(this.get()));
+      return f;
+    }
+
+    pragma "no doc"
+    proc acquire(newRef: FutureClass) {
+      classRef = newRef;
+      classRef.incRefCount();
+    }
+
+    pragma "no doc"
+    proc acquire() {
+      classRef.incRefCount();
+    }
+
+    pragma "no doc"
+    proc release() {
+      var rc = classRef.decRefCount();
+      if rc == 1 {
+        delete classRef;
+        classRef = nil;
+      }
+    }
+
+  } // record Future
+
+  pragma "no doc"
+  pragma "init copy fn"
+  proc chpl__initCopy(x: Future) {
+    x.acquire();
+    return x;
+  }
+
+  pragma "no doc"
+  pragma "auto copy fn"
+  proc chpl__autoCopy(x: Future) {
+    x.acquire();
+    return x;
+  }
+
+  pragma "no doc"
+  proc =(ref lhs: Future, rhs: Future) {
+  if lhs.classRef != nil then
```

Comment:
I take that last part back... I see now that lhs.acquire(value) assigns the field.  Perhaps that is sufficient to guarantee no deadlock cycles, but I'd be interested to know if you think that's the case. Something like a deadlock like this (where it probably has syntax errors):
```
a: future;
b: future;
a = async( lamda() { b.get(); } );
b = async( lambda() { a.get(); } );
```

Is it the case that such examples don't deadlock because the *class instance* itself is immutable, and we're just changing what class instance a or b point to?